include/zephyr/drivers/adc.h

```diff
@@ -731,6 +706,118 @@ static inline uint16_t adc_ref_internal(const struct device *dev)
 	return api->ref_internal;
 }
 
+/**
+ * @brief Convert a raw ADC value to millivolts.
+ *
+ * This function performs the necessary conversion to transform a raw
+ * ADC measurement to a voltage in millivolts.
+ *
+ * @param ref_mv the reference voltage used for the measurement, in
+ * millivolts.  This may be from adc_ref_internal() or a known
+ * external reference.
+ *
+ * @param gain the ADC gain configuration used to sample the input
+ *
+ * @param resolution the number of bits in the absolute value of the
+ * sample.  For differential sampling this may be one less than the
+ * resolution in struct adc_sequence.
+ *
+ * @param valp pointer to the raw measurement value on input, and the
+ * corresponding millivolt value on successful conversion.  If
+ * conversion fails the stored value is left unchanged.
+ *
+ * @retval 0 on successful conversion
+ * @retval -EINVAL if the gain is not reversible
+ */
+static inline int adc_raw_to_millivolts(int32_t ref_mv,
+					enum adc_gain gain,
+					uint8_t resolution,
+					int32_t *valp)
+{
+	int32_t adc_mv = *valp * ref_mv;
+	int ret = adc_gain_invert(gain, &adc_mv);
+
+	if (ret == 0) {
+		*valp = (adc_mv >> resolution);
+	}
+
+	return ret;
+}
+
+/**
+ * @brief Convert a raw ADC value to millivolts using information stored
+ * in a struct adc_dt_spec.
+ *
+ * @param[in] spec ADC specification from Devicetree.
+ * @param[in,out] valp Pointer to the raw measurement value on input, and the
+ *                     corresponding millivolt value on successful conversion.
+ *                     If conversion fails the stored value is left unchanged.
+ *
+ * @return A value from adc_raw_to_millivolts().
+ * @see adc_raw_to_millivolts()
+ */
+static inline int adc_raw_to_millivolts_dt(const struct adc_dt_spec *spec,
+					   int32_t *valp)
+{
+	int32_t vref_mv;
+	uint8_t resolution;
+
+	if (!spec->channel_cfg_dt_node_exists) {
+		return -ENOTSUP;
+	}
+
+	if (spec->channel_cfg.reference == ADC_REF_INTERNAL) {
+		vref_mv = (int32_t)adc_ref_internal(spec->dev);
+	} else {
+		vref_mv = spec->vref_mv;
+	}
+
+	resolution = spec->resolution;
+
+	/*
+	 * For differential channels, one bit less needs to be specified
+	 * for resolution to achieve correct conversion.
+	 */
+	if (spec->channel_cfg.differential) {
+		resolution -= 1U;
+	}
```

Comment:
This not match with the documentation at lines 722~723. I mean, it "`needs to be`" or it "`may be`". Looking the code it seems that `needs to be` is the correct one.

Reply:
@anangl could you comment on that? From an API perspective it's marked as "optional", but sample did this operation unconditionally.

Reply:
In my opinion this "may be" in the description of `adc_raw_to_millivolts` should be corrected, as it is not consistent with the description of `resolution`: https://github.com/zephyrproject-rtos/zephyr/blob/2c1ef5598cb809d665fbe4bf496725a33a0928ea/include/zephyr/drivers/adc.h#L519-L526 This -1 cannot be "optional".